### PR TITLE
Stablize instrumentation.js

### PR DESCRIPTION
**Pull request by huozhi**
### What

- After adding a new API `onRequestError()` and a few fixes on `register()` for `instrumentation.js`, it's time to promote it as stable API.
- Add warning `"experimental.instrumentationHook is no longer needed to be configured in Next.js"` when you configure a `instrumentationHook` option in next.config.js

Related PRs
#67539, #67671, #67703, #67848, #67859, #67856, #67805, #68616, #68672, #68764, #68845 #68976, #68983

Closes NDX-89

### CODE REVIEW

**Comment by huozhi** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/vercel/next.js/68853?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#68853** <a href="https://app.graphite.dev/github/pr/vercel/next.js/68853?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#68845** <a href="https://app.graphite.dev/github/pr/vercel/next.js/68845?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @huozhi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| buildDuration | 17.6s | 16s | N/A |
| buildDurationCached | 9.4s | 7.9s | N/A |
| nodeModulesSize | 356 MB | 356 MB | ⚠️ +19.7 kB |
| nextStartRea..uration (ms) | 426ms | 430ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| 4958.HASH.js gzip | 169 B | 168 B | N/A |
| 6405-HASH.js gzip | 5.19 kB | 5.19 kB | N/A |
| 8711-HASH.js gzip | 37.9 kB | 37.9 kB | N/A |
| d09468e9-HASH.js gzip | 51.9 kB | 51.9 kB | N/A |
| framework-HASH.js gzip | 56.8 kB | 56.7 kB | N/A |
| main-app-HASH.js gzip | 225 B | 225 B | ✓ |
| main-HASH.js gzip | 32.5 kB | 32.5 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 225 B | 225 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 192 B | N/A |
| _error-HASH.js gzip | 191 B | 192 B | N/A |
| amp-HASH.js gzip | 512 B | 511 B | N/A |
| css-HASH.js gzip | 342 B | 344 B | N/A |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 266 B | N/A |
| head-HASH.js gzip | 361 B | 364 B | N/A |
| hooks-HASH.js gzip | 390 B | 393 B | N/A |
| image-HASH.js gzip | 4.4 kB | 4.4 kB | N/A |
| index-HASH.js gzip | 268 B | 267 B | N/A |
| link-HASH.js gzip | 2.82 kB | 2.81 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 324 B | 324 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.16 kB | 1.16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| _buildManifest.js gzip | 747 B | 749 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| index.html gzip | 522 B | 521 B | N/A |
| link.html gzip | 536 B | 537 B | N/A |
| withRouter.html gzip | 518 B | 518 B | ✓ |
| Overall change | 518 B | 518 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| edge-ssr.js gzip | 127 kB | 127 kB | N/A |
| page.js gzip | 173 kB | 173 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 668 B | 670 B | N/A |
| middleware-r..fest.js gzip | 156 B | 156 B | ✓ |
| middleware.js gzip | 29.6 kB | 29.6 kB | ✓ |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 30.6 kB | 30.6 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| 928-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 928.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 307 kB | 307 kB | N/A |
| app-page-exp..prod.js gzip | 121 kB | 121 kB | N/A |
| app-page-tur..prod.js gzip | 133 kB | 133 kB | N/A |
| app-page-tur..prod.js gzip | 129 kB | 129 kB | N/A |
| app-page.run...dev.js gzip | 297 kB | 297 kB | N/A |
| app-page.run..prod.js gzip | 117 kB | 117 kB | N/A |
| app-route-ex...dev.js gzip | 30.7 kB | 30.7 kB | ✓ |
| app-route-ex..prod.js gzip | 20.7 kB | 20.7 kB | ✓ |
| app-route-tu..prod.js gzip | 20.7 kB | 20.7 kB | ✓ |
| app-route-tu..prod.js gzip | 20.5 kB | 20.5 kB | ✓ |
| app-route.ru...dev.js gzip | 32.3 kB | 32.3 kB | ✓ |
| app-route.ru..prod.js gzip | 20.5 kB | 20.5 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.6 kB | 9.6 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.4 kB | 11.4 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.59 kB | 9.59 kB | ✓ |
| pages-turbo...prod.js gzip | 21.6 kB | 21.6 kB | ✓ |
| pages.runtim...dev.js gzip | 27.5 kB | 27.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.6 kB | 21.6 kB | ✓ |
| server.runti..prod.js gzip | 56.9 kB | 56.9 kB | N/A |
| Overall change | 247 kB | 247 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 08-13-stable_instrumentation | Change |
| - | - | - | - |
| 0.pack gzip | 1.49 MB | 1.49 MB | ⚠️ +294 B |
| index.pack gzip | 127 kB | 126 kB | N/A |
| Overall change | 1.49 MB | 1.49 MB | ⚠️ +294 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 9842: /***/ (
+    /***/ 5684: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(8722);
+          return __webpack_require__(2665);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 5040: /***/ (module, exports, __webpack_require__) => {
+    /***/ 3608: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,17 +40,17 @@
         __webpack_require__(2957)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(5187)
+        __webpack_require__(7569)
       );
-      const _getimgprops = __webpack_require__(1332);
-      const _imageconfig = __webpack_require__(3384);
-      const _imageconfigcontextsharedruntime = __webpack_require__(8875);
-      const _warnonce = __webpack_require__(4516);
-      const _routercontextsharedruntime = __webpack_require__(5002);
+      const _getimgprops = __webpack_require__(4429);
+      const _imageconfig = __webpack_require__(2950);
+      const _imageconfigcontextsharedruntime = __webpack_require__(6587);
+      const _warnonce = __webpack_require__(8719);
+      const _routercontextsharedruntime = __webpack_require__(2062);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(5145)
+        __webpack_require__(232)
       );
-      const _usemergedref = __webpack_require__(4926);
+      const _usemergedref = __webpack_require__(9935);
       // This is replaced by webpack define plugin
       const configEnv = {
         deviceSizes: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
@@ -371,7 +371,7 @@
       /***/
     },
 
-    /***/ 4926: /***/ (module, exports, __webpack_require__) => {
+    /***/ 9935: /***/ (module, exports, __webpack_require__) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -440,7 +440,7 @@
       /***/
     },
 
-    /***/ 1332: /***/ (
+    /***/ 4429: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -456,9 +456,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(4516);
-      const _imageblursvg = __webpack_require__(1569);
-      const _imageconfig = __webpack_require__(3384);
+      const _warnonce = __webpack_require__(8719);
+      const _imageblursvg = __webpack_require__(3207);
+      const _imageconfig = __webpack_require__(2950);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -830,7 +830,7 @@
       /***/
     },
 
-    /***/ 1569: /***/ (__unused_webpack_module, exports) => {
+    /***/ 3207: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -885,7 +885,7 @@
       /***/
     },
 
-    /***/ 1674: /***/ (
+    /***/ 3455: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -912,10 +912,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(4345);
-      const _getimgprops = __webpack_require__(1332);
-      const _imagecomponent = __webpack_require__(5040);
+      const _getimgprops = __webpack_require__(4429);
+      const _imagecomponent = __webpack_require__(3608);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(5145)
+        __webpack_require__(232)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -947,7 +947,7 @@
       /***/
     },
 
-    /***/ 5145: /***/ (__unused_webpack_module, exports) => {
+    /***/ 232: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -982,7 +982,7 @@
       /***/
     },
 
-    /***/ 8722: /***/ (
+    /***/ 2665: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -999,8 +999,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@19.0.0-rc-187dd6a7-20240806/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(8548);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/next@file+..+main-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-187dd6a7-20240806_re_6tugpsz4gl2xsqrwds5bwcbi3y/node_modules/next/image.js
-      var next_image = __webpack_require__(6629);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/next@file+..+diff-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-187dd6a7-20240806_re_wxf3je55uhqnotjiz3wkaonabu/node_modules/next/image.js
+      var next_image = __webpack_require__(2856);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -1030,12 +1030,12 @@
       /***/
     },
 
-    /***/ 6629: /***/ (
+    /***/ 2856: /***/ (
       module,
       __unused_webpack_exports,
       __webpack_require__
     ) => {
-      module.exports = __webpack_require__(1674);
+      module.exports = __webpack_require__(3455);
 
       /***/
     },
@@ -1045,7 +1045,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(9842)
+      __webpack_exec__(5684)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>8711-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5859263a6eedc9f4dcc6180c0369423a7f9c3326

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 5c761beb3bdb167bbe8e07faadfb9463b0dbe2a4

`pnpm test-dev test/e2e/app-dir/navigation/navigation.test.ts`
- app dir - navigation > scroll restoration > should restore original scroll position when navigating back

<details>
<summary>Expand output</summary>

  ● app dir - navigation › scroll restoration › should restore original scroll position when navigating back

    page.waitForSelector: Timeout 60000ms exceeded.
    Call log:
      - waiting for locator('#back-button')

      423 |     return this.chain(() => {
      424 |       return page
    > 425 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      426 |         .then(async (el) => {
      427 |           // it seems selenium waits longer and tests rely on this behavior
      428 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:425:10)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/integration/config-experimental-warning/test/index.test.js [hunk 1]
```diff
@@ -139,7 +139,7 @@ describe('Config Experimental Warning', () => {
           experimental: {
             workerThreads: true,
             scrollRestoration: true,
-            instrumentationHook: true,
+            parallelServerCompiles: true,
```
Comment (wyattjoh) [inline test/integration/config-experimental-warning/test/index.test.js:142]: Assuming this is unrelated, just adding more test params?

Comment (huozhi) [reply test/integration/config-experimental-warning/test/index.test.js:142]: Since `instrumentationHook` is removed, so I got another exp flag here, just for testing the output logs. see the checks below.

File: packages/next/src/server/config.ts [hunk 2]
```diff
@@ -471,6 +471,13 @@ function assignDefaults(
     silent
   )
 
+  warnOptionHasBeenDeprecated(
```
Comment (ztanner) [inline packages/next/src/server/config.ts:474]: should this be `warnOptionHasBeenMovedOutOfExperimental`?

File: test/integration/config-experimental-warning/test/index.test.js (hunk 1 above)
Comment (ztanner) [reply test/integration/config-experimental-warning/test/index.test.js:142]: I think it's fine to just remove the assertion without adding a new experimental flag

Comment (huozhi) [reply test/integration/config-experimental-warning/test/index.test.js:142]: We're logging all the experimentals in build mode, it's different between dev and build. In dev mode it only logs 3 maximum. This case is to test in build mode we'll log all

File: packages/next/src/server/config.ts (hunk 2 above)
Comment (huozhi) [reply packages/next/src/server/config.ts:474]: I made a new one cause `warnOptionHasBeenMovedOutOfExperimental` will say the flag is moved to xxx, mostly for the flags that are still remained. Here we delete the flag, similar others like `swcMinify`. So we add a warning for users that they can delete the unexisted flag

**Review by ztanner — APPROVED** [review]
